src/protocol/boolean/saturating_sum.rs

```diff
@@ -0,0 +1,170 @@
+use crate::{
+    error::Error,
+    ff::Gf2,
+    protocol::{context::Context, step::BitOpStep, BasicProtocols, RecordId},
+    secret_sharing::{BitDecomposed, Linear as LinearSecretSharing},
+};
+
+#[derive(Debug)]
+pub struct SaturatingSum<S: LinearSecretSharing<Gf2>> {
+    pub sum: BitDecomposed<S>,
+    pub is_saturated: S,
+}
+
+impl<S: LinearSecretSharing<Gf2>> SaturatingSum<S> {
+    pub fn new(value: BitDecomposed<S>, is_saturated: S) -> SaturatingSum<S> {
```

Comment:
Is there ever a case where you would construct one of these with is_saturated set to anything other than ZERO?  I ask, because it seems like you might want to avoid the case where `sum` is less than the maximum, say:

```rust
   let x = SaturatingSum::new(get_bits::<Gf2>(20, 5), ctx.share_known_value(1));
```

It's generally not a good idea to allow objects to be constructed in a way that allows invariants like this to be violated.  I can't see a concrete problem in this case, but there you go.

Reply:
The only case would probably be in tests, but I'm not currently using that.

As I note in my comment above, it's not an invariant violation to construct a SaturatingSum where the lower bits are random stuff and the "is_saturated" is a 1. That is just interpreted as being saturated.

Reply:
OK, so there is one place that I call it with ONE, and that's in the `add()` method itself. It returns a new SaturatingSum, and it might be saturated.

Reply:
It seems like you might have separate `new` and `new_saturated` functions for that then.